File: src/commands/stats.ts
import Discord, { SlashCommandBuilder, AttachmentBuilder } from 'discord.js'
import { database } from '../helpers/database'
import { config } from '../config'
import { getStreak } from '../helpers/streaks'
import { channelGuard } from '../helpers/guards'
import { get_data, make_chart, get_all_time } from '../helpers/stats'

export = {
  data: new SlashCommandBuilder()
    .setName('stats')
    .setDescription('Gets your stats or the stats of a specified user.')
    .addSubcommand(subcommand =>
      subcommand.setName('user')
        .setDescription('Gets your stats, or the stats of a specified user.')
        .addUserOption(option =>
          option.setName('user')
            .setDescription('The user to get the stats of.'))
        .addStringOption(option =>
          option.setName('type')
            .setDescription('The type of stats to get.')
            .addChoices(
              { name: 'Meditation Minutes', value: 'meditation_minutes' },
              { name: 'Meditation Count', value: 'meditation_count' }
            )
            .setRequired(false))
        .addStringOption(option =>
          option.setName('timeframe')
            .setDescription('The timeframe to get the stats for.')
            .addChoices(
              { name: 'Yearly', value: 'yearly' },
              { name: 'Monthly', value: 'monthly' },
              { name: 'Weekly', value: 'weekly' },
              { name: 'Daily', value: 'daily' }
            )
            .setRequired(false)))
    .addSubcommand(subcommand =>
      subcommand.setName('server')
        .setDescription('Gets the stats of the current guild.')
        .addStringOption(option =>
          option.setName('type')
            .setDescription('The type of stats to get.')
            .addChoices(
              { name: 'Meditation Minutes', value: 'meditation_minutes' },
              { name: 'Meditation Count', value: 'meditation_count' }
            )
            .setRequired(false))
        .addStringOption(option =>
          option.setName('timeframe')
            .setDescription('The timeframe to get the stats for.')
            .addChoices(
              { name: 'Yearly', value: 'yearly' },
              { name: 'Monthly', value: 'monthly' },
              { name: 'Weekly', value: 'weekly' },
              { name: 'Daily', value: 'daily' }
            )
            .setRequired(false)))
    .setDMPermission(false),
  async execute (interaction) {
    if (!(await channelGuard)(interaction, [config.channels.meditation, config.channels.commands], interaction.channelId)) return

    const subcommand = interaction.options.getSubcommand()

    if (subcommand === 'user') {
      const user = interaction.options.getUser('user') || interaction.user
      const type = interaction.options.getString('type') || 'meditation_minutes'
      const timeframe = interaction.options.getString('timeframe') || 'daily'

      const member = interaction.guild.members.cache.get(user.id)

      const recent_meditations = await database.meditations.findMany({
        where: {
          session_user: user.id,
          session_guild: interaction.guild.id
        },
        orderBy: [
          {
            occurred_at: 'desc'
          }
        ],
        take: 3
      })

      const meditation_aggregation = await database.meditations.aggregate({
        where: {
          session_user: user.id,
          session_guild: interaction.guild.id
        },
        _count: {
          id: true
        },
        _sum: {
          session_time: true
        }
      })

      const user_count = meditation_aggregation._count.id || 0
      const user_time = meditation_aggregation._sum.session_time || 0
      const streak = await getStreak(interaction.client, interaction.guild, user)

      const meditations: string[] = []

      recent_meditations.forEach((meditation) => {
        const date = new Date(meditation.occurred_at)
        const month = date.getUTCMonth() + 1
        const day = date.getUTCDate()
        const year = date.getUTCFullYear()

        meditations.push(
          `**${meditation.session_time}m** on ${day}/${month}/${year}\nID: \`${meditation.id}\``
        )
      })

      // Gets the data for the chart
      const raw_data: any = await get_data(timeframe, interaction.guild.id, user.id)

      const embedColor = member.displayHexColor === '#000000' ? '#ffffff' : member.displayHexColor
      const canvas = make_chart(raw_data, timeframe, type, embedColor)

      const timeframeWords = {
        daily: 'Days',
        weekly: 'Weeks',
        monthly: 'Months',
        yearly: 'Years'
      }

      const fields = [
        {
          name: 'All-Time Meditation Minutes',
          value: `\`\`\`${user_time.toLocaleString()}\`\`\``,
          inline: false
        },
        {
          name: 'All-Time Session Count',
          value: `\`\`\`${user_count.toLocaleString()}\`\`\``,
          inline: false
        },
        { name: `Minutes The Past 12 ${timeframeWords[timeframe]}`, value: `\`\`\`${raw_data.reduce((a, b) => a + Number(b.total_time), 0).toLocaleString()}\`\`\``, inline: true },
        { name: `Sessions The Past 12 ${timeframeWords[timeframe]}`, value: `\`\`\`${raw_data.reduce((a, b) => a + Number(b.count), 0).toLocaleString()}\`\`\``, inline: true },
        {
          name: 'Current Streak',
          value: `\`\`\`${streak} days\`\`\``,
          inline: false
        }
      ]

      const attachment = new AttachmentBuilder(canvas, { name: 'chart.png' })

      const statsEmbed = new Discord.EmbedBuilder()
        .setColor(config.embedColor)
        .setAuthor({ name: `${user.username}'s Stats`, iconURL: user.avatarURL() })
        .addFields(...fields)
        .setImage('attachment://chart.png')

      return interaction.reply({
        embeds: [statsEmbed],
        files: [attachment]
      })
    } else if (subcommand === 'server') {
      const type = interaction.options.getString('type') || 'meditation_count'
      const timeframe = interaction.options.getString('timeframe') || 'daily'

      // Gets the data for the chart
      const raw_data: any = await get_data(timeframe, interaction.guild.id)

      console.log(timeframe)
      console.log(raw_data)

      const canvas = make_chart(raw_data, timeframe, type)

      const all_time = await get_all_time(interaction)
      const guild_name = interaction.guild.name

      const timeframeWords = {
        daily: 'Days',
        weekly: 'Weeks',
        monthly: 'Months',
        yearly: 'Years'
      }

      const allSessionTime = all_time._sum.session_time ? all_time._sum.session_time : 0
      const allSessionCount = all_time._count.id ? all_time._count.id : 0

      const attachment = new AttachmentBuilder(canvas, { name: 'chart.png' })
      const embed = new Discord.EmbedBuilder()
        .setTitle('Stats')
        .setColor(config.embedColor)
        .setDescription(`Here are the stats for **${guild_name}**.`)
        .addFields(
          { name: 'All-Time Meditation Minutes', value: `\`\`\`${allSessionTime.toLocaleString()}\`\`\`` },
          { name: 'All-Time Session Count', value: `\`\`\`${allSessionCount.toLocaleString()}\`\`\`` },
          { name: `Minutes The Past 12 ${timeframeWords[timeframe]}`, value: `\`\`\`${raw_data.reduce((a, b) => a + Number(b.total_time), 0).toLocaleString()}\`\`\``, inline: true },
          { name: `Sessions The Past 12 ${timeframeWords[timeframe]}`, value: `\`\`\`${raw_data.reduce((a, b) => a + Number(b.count), 0).toLocaleString()}\`\`\``, inline: true }
        )
        .setImage('attachment://chart.png')
      return interaction.reply({ embeds: [embed], files: [attachment] })
    }

    return interaction.reply({ content: 'Please specify a subcommand.', ephemeral: true })
  }
}

File: src/commands/stats.rs
use crate::config::{BloomBotEmbed, ROLES};
use crate::database::Timeframe;
use crate::database::{DatabaseHandler, TrackingProfile};
use crate::Context;
use crate::{charts, config};
use anyhow::Result;
use poise::serenity_prelude::{self as serenity, builder::*};
use poise::ChoiceParameter;

#[derive(poise::ChoiceParameter)]
pub enum StatsType {
  #[name = "Minutes"]
  MeditationMinutes,
  #[name = "Count"]
  MeditationCount,
}

#[derive(poise::ChoiceParameter)]
pub enum Privacy {
  #[name = "Private"]
  Private,
  #[name = "Public"]
  Public,
}

#[derive(poise::ChoiceParameter)]
pub enum Theme {
  #[name = "Light Mode"]
  LightMode,
  #[name = "Dark Mode"]
  DarkMode,
}

/// Show stats for a user or the server
///
/// Shows stats for yourself, a specified user, or the whole server.
#[poise::command(
  slash_command,
  category = "Meditation Tracking",
  subcommands("user", "server"),
  subcommand_required,
  guild_only
)]
pub async fn stats(_: Context<'_>) -> Result<()> {
  Ok(())
}

/// Show stats for a user
///
/// Shows stats for yourself or a specified user.
///
/// Defaults to daily minutes for yourself. Optionally specify the user, type (minutes or session count), and/or timeframe (daily, weekly, monthly, or yearly).
#[poise::command(slash_command)]
pub async fn user(
  ctx: Context<'_>,
  #[description = "The user to get the stats of (Defaults to you)"] user: Option<serenity::User>,
  #[description = "The type of stats to get (Defaults to minutes)"]
  #[rename = "type"]
  stats_type: Option<StatsType>,
  #[description = "The timeframe to get the stats for (Defaults to daily)"] timeframe: Option<
    Timeframe,
  >,
  #[description = "Set visibility of response (Defaults to public)"] privacy: Option<Privacy>,
  #[description = "Toggle between light mode and dark mode (Defaults to dark mode)"] theme: Option<
    Theme,
  >,
) -> Result<()> {
  let data = ctx.data();
  let mut transaction = data.db.start_transaction_with_retry(5).await?;

  let guild_id = ctx.guild_id().unwrap();

  let user = user.unwrap_or_else(|| ctx.author().clone());
  let user_nick_or_name = match user.nick_in(&ctx, guild_id).await {
    Some(nick) => nick,
    None => user.name.clone(),
  };

  let tracking_profile =
    match DatabaseHandler::get_tracking_profile(&mut transaction, &guild_id, &user.id).await? {
      Some(tracking_profile) => tracking_profile,
      None => TrackingProfile {
        ..Default::default()
      },
    };

  let privacy = match privacy {
    Some(privacy) => match privacy {
      Privacy::Private => true,
      Privacy::Public => false,
    },
    None => tracking_profile.stats_private,
  };

  if privacy {
    ctx.defer_ephemeral().await?;
  } else {
    ctx.defer().await?;
  }

  if ctx.author().id != user.id
    && tracking_profile.stats_private
    && !ctx.author().has_role(&ctx, guild_id, ROLES.staff).await?
  {
    ctx
      .send(
        poise::CreateReply::default()
          .content(format!(
            "Sorry, {}'s stats are set to private.",
            user_nick_or_name
          ))
          .ephemeral(true)
          .allowed_mentions(serenity::CreateAllowedMentions::new()),
      )
      .await?;

    return Ok(());
  }

  let stats_type = stats_type.unwrap_or(StatsType::MeditationMinutes);
  let timeframe = timeframe.unwrap_or(Timeframe::Daily);

  let timeframe_header = match timeframe {
    Timeframe::Yearly => "Years",
    Timeframe::Monthly => "Months",
    Timeframe::Weekly => "Weeks",
    Timeframe::Daily => "Days",
  };

  let stats =
    DatabaseHandler::get_user_stats(&mut transaction, &guild_id, &user.id, &timeframe).await?;

  let mut embed = BloomBotEmbed::new();
  embed = embed
    .title(format!("Stats for {}", user_nick_or_name))
    .author(CreateEmbedAuthor::new(format!("{}'s Stats", user_nick_or_name)).icon_url(user.face()));

  match stats_type {
    StatsType::MeditationMinutes => {
      embed = embed
        .field(
          "All-Time Meditation Minutes",
          format!("```{}```", stats.all_minutes),
          true,
        )
        .field(
          format!("Minutes The Past 12 {}", timeframe_header),
          format!("```{}```", stats.timeframe_stats.sum.unwrap_or(0)),
          true,
        );
    }
    StatsType::MeditationCount => {
      embed = embed
        .field(
          "All-Time Session Count",
          format!("```{}```", stats.all_count),
          true,
        )
        .field(
          format!("Sessions The Past 12 {}", timeframe_header),
          format!("```{}```", stats.timeframe_stats.count.unwrap_or(0)),
          true,
        );
    }
  }

  // Role-based bar color for donators; default otherwise
  let bar_color = if user.has_role(&ctx, guild_id, config::ROLES.patreon).await?
    || user.has_role(&ctx, guild_id, config::ROLES.kofi).await?
  {
    match guild_id.member(&ctx, user.id).await?.colour(&ctx) {
      Some(color) => (color.r(), color.g(), color.b(), 1.0),
      None => (253, 172, 46, 1.0),
    }
  } else {
    (253, 172, 46, 1.0)
  };

  // Role-based bar color for all users
  //let bar_color = match guild_id.member(&ctx, user.id).await?.colour(&ctx) {
  //  Some(color) => (color.r(), color.g(), color.b(), 1.0),
  //  None => (253, 172, 46, 1.0)
  //};

  let light_mode = match theme {
    Some(theme) => match theme {
      Theme::LightMode => true,
      Theme::DarkMode => false,
    },
    None => false,
  };

  let chart_stats =
    DatabaseHandler::get_user_chart_stats(&mut transaction, &guild_id, &user.id, &timeframe)
      .await?;
  let chart_drawer = charts::ChartDrawer::new()?;
  let chart = chart_drawer
    .draw(&chart_stats, &timeframe, &stats_type, bar_color, light_mode)
    .await?;
  let file_path = chart.get_file_path();

  embed = embed.image(chart.get_attachment_url());

  let average = match stats_type {
    StatsType::MeditationMinutes => stats.timeframe_stats.sum.unwrap_or(0) / 12,
    StatsType::MeditationCount => stats.timeframe_stats.count.unwrap_or(0) / 12,
  };

  let stats_type_label = match stats_type {
    StatsType::MeditationMinutes => "minutes",
    StatsType::MeditationCount => "sessions",
  };

  //Hide streak in footer if streaks disabled
  if tracking_profile.streaks_active {
    embed = embed.footer(CreateEmbedFooter::new(format!(
      "Avg. {} {}: {}・Current streak: {}",
      timeframe.name().to_lowercase(),
      stats_type_label,
      average,
      stats.streak
    )));
  } else {
    embed = embed.footer(CreateEmbedFooter::new(format!(
      "Average {} {}: {}",
      timeframe.name().to_lowercase(),
      stats_type_label,
      average
    )));
  }

  ctx
    .send({
      let mut f =
        poise::CreateReply::default().attachment(CreateAttachment::path(&file_path).await?);
      f.embeds = vec![embed.to_owned()];

      f
    })
    .await?;

  Ok(())
}

/// Show stats for the server
///
/// Shows stats for the whole server.
///
/// Defaults to daily minutes. Optionally specify the type (minutes or session count) and/or timeframe (daily, weekly, monthly, or yearly).
#[poise::command(slash_command)]
pub async fn server(
  ctx: Context<'_>,
  #[description = "The type of stats to get (Defaults to minutes)"] stats_type: Option<StatsType>,
  #[description = "The timeframe to get the stats for (Defaults to daily)"] timeframe: Option<
    Timeframe,
  >,
  #[description = "Toggle between light mode and dark mode (Defaults to dark mode)"] theme: Option<
    Theme,
  >,
) -> Result<()> {
  ctx.defer().await?;

  let data = ctx.data();

  let guild_id = ctx.guild_id().unwrap();
  let guild_name = guild_id.name(ctx).unwrap();

  let stats_type = stats_type.unwrap_or(StatsType::MeditationMinutes);
  let timeframe = timeframe.unwrap_or(Timeframe::Daily);

  let timeframe_header = match timeframe {
    Timeframe::Yearly => "Years",
    Timeframe::Monthly => "Months",
    Timeframe::Weekly => "Weeks",
    Timeframe::Daily => "Days",
  };

  let mut transaction = data.db.start_transaction_with_retry(5).await?;

  let stats = DatabaseHandler::get_guild_stats(&mut transaction, &guild_id, &timeframe).await?;

  let mut embed = BloomBotEmbed::new();
  embed = embed.title(format!("Stats for {}", guild_name)).author(
    CreateEmbedAuthor::new(format!("{}'s Stats", guild_name))
      .icon_url(ctx.guild().unwrap().icon_url().unwrap_or_default()),
  );

  match stats_type {
    StatsType::MeditationMinutes => {
      embed = embed
        .field(
          "All-Time Meditation Minutes",
          format!("```{}```", stats.all_minutes),
          true,
        )
        .field(
          format!("Minutes The Past 12 {}", timeframe_header),
          format!("```{}```", stats.timeframe_stats.sum.unwrap_or(0)),
          true,
        );
    }
    StatsType::MeditationCount => {
      embed = embed
        .field(
          "All-Time Session Count",
          format!("```{}```", stats.all_count),
          true,
        )
        .field(
          format!("Sessions The Past 12 {}", timeframe_header),
          format!("```{}```", stats.timeframe_stats.count.unwrap_or(0)),
          true,
        );
    }
  }

  let bar_color = (253, 172, 46, 1.0);
  let light_mode = match theme {
    Some(theme) => match theme {
      Theme::LightMode => true,
      Theme::DarkMode => false,
    },
    None => false,
  };

  let chart_stats =
    DatabaseHandler::get_guild_chart_stats(&mut transaction, &guild_id, &timeframe).await?;
  let chart_drawer = charts::ChartDrawer::new()?;
  let chart = chart_drawer
    .draw(&chart_stats, &timeframe, &stats_type, bar_color, light_mode)
    .await?;
  let file_path = chart.get_file_path();

  embed = embed.image(chart.get_attachment_url());

  ctx
    .send({
      let mut f =
        poise::CreateReply::default().attachment(CreateAttachment::path(&file_path).await?);
      f.embeds = vec![embed.to_owned()];

      f
    })
    .await?;

  Ok(())
}
